File: SeoulWomen/app/src/main/java/kr/susemi99/seoulwomen/models/RowItem.java
package kr.susemi99.seoulwomen.models;

import android.text.TextUtils;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;

import kr.susemi99.seoulwomen.utils.DateFormatter;
import kr.susemi99.seoulwomen.utils.StringUtil;

public class RowItem {
  @SerializedName("CLASS_CODE")
  public String classCode;
  @SerializedName("CLASS_NAME")
  public String className;
  @SerializedName("ORGAN_CODE")
  public String organCode;
  @SerializedName("ORGAN_NAME")
  public String organName;
  @SerializedName("DIFFICULTY")
  public String difficulty;
  @SerializedName("DIFFICULTY_NAME")
  public String difficultyName;
  @SerializedName("RECEIVE_FROM")
  public String receiveFrom;
  @SerializedName("RECEIVE_TO")
  public String receiveTo;
  @SerializedName("RECEIVE_TIME_FROM")
  public String receiveTimeFrom;
  @SerializedName("RECEIVE_TIME_TO")
  public String receiveTimeTo;
  @SerializedName("EDUCATE_FROM")
  public String educateFrom;
  @SerializedName("EDUCATE_TO")
  public String educateTo;
  @SerializedName("EDUCATE_TIME_FROM")
  public String educateTimeFrom;
  @SerializedName("EDUCATE_TIME_TO")
  public String educateTimeTo;
  @SerializedName("MONDAY")
  public String monday;
  @SerializedName("TUESDAY")
  public String tuesday;
  @SerializedName("WEDNESDAY")
  public String wednesday;
  @SerializedName("THURSDAY")
  public String thursday;
  @SerializedName("FRIDAY")
  public String friday;
  @SerializedName("SATURDAY")
  public String saturday;
  @SerializedName("SUNDAY")
  public String sunday;
  @SerializedName("COLLECT_NUM")
  public String collectNum;
  @SerializedName("SPARE_NUM")
  public String spareNum;
  @SerializedName("EDUCATE_FEE")
  public String educateFee;
  @SerializedName("VISIT_RECEIVE_FLAG")
  public String visitReceiveFlag;
  @SerializedName("ONLINE_RECEIVE_FLAG")
  public String onlineReceiveFlag;
  @SerializedName("URL")
  public String url;

  public String displayReceiveFrom() {
    return DateFormatter.parseDate(receiveFrom, "yyyyMMdd");
  }

  public String displayReceiveTo() {
    return DateFormatter.parseDate(receiveTo, "yyyyMMdd");
  }

  public String displayEducateFrom() {
    return DateFormatter.parseDate(educateFrom, "yyyyMMdd");
  }

  public String displayEducateTo() {
    return DateFormatter.parseDate(educateTo, "yyyyMMdd");
  }

  public String displaySpareNum() {
    return StringUtil.addComma(Float.parseFloat(spareNum));
  }

  public String displayDays() {
    StringBuilder builder = new StringBuilder();
    if(!TextUtils.isEmpty(monday)) {
      builder.append("월");
    }
    if(!TextUtils.isEmpty(tuesday)) {
      builder.append("화");
    }
    if(!TextUtils.isEmpty(wednesday)) {
      builder.append("수");
    }
    if(!TextUtils.isEmpty(thursday)) {
      builder.append("목");
    }
    if(!TextUtils.isEmpty(friday)) {
      builder.append("금");
    }
    if(!TextUtils.isEmpty(saturday)) {
      builder.append("토");
    }
    if(!TextUtils.isEmpty(sunday)) {
      builder.append("일");
    }
    return builder.toString();
  }

  public String displayCollectNum() {
    return StringUtil.addComma(Float.parseFloat(collectNum));
  }

  public String displayEducateFee() {
    float floatFee = Float.parseFloat(educateFee);
    int intFee = (int) floatFee;
    if(intFee == 0) {
      return "무료";
    } else {
      return StringUtil.addComma(intFee) + "원";
    }
  }

  public String displayHowToRegist() {
    ArrayList<String> result = new ArrayList<>();
    if(visitReceiveFlag.equals("Y")) {
      result.add("방문");
    }
    if(onlineReceiveFlag.equals("Y")) {
      result.add("온라인");
    }
    return TextUtils.join(", ", result);
  }
}

File: SeoulWomen/app/src/main/java/kr/susemi99/seoulwomen/model/RowItem.kt
package kr.susemi99.seoulwomen.model

import kotlinx.serialization.Contextual
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kr.susemi99.seoulwomen.extension.dayName
import kr.susemi99.seoulwomen.util.serializer.DecimalIntSerializer
import java.time.LocalDate
import java.time.LocalTime

@Serializable
data class RowItem(
  @SerialName("CLASS_CODE") val classCode: String,
  @SerialName("CLASS_NAME") val className: String,
  @SerialName("ORGAN_CODE") val organCode: String,
  @SerialName("ORGAN_NAME") val organName: String,
  @SerialName("DIFFICULTY") val difficultyCode: String,
  @SerialName("DIFFICULTY_NAME") val difficultyName: String,
  @Contextual @SerialName("RECEIVE_FROM") val receiveFrom: LocalDate,
  @Contextual @SerialName("RECEIVE_TO") val receiveTo: LocalDate,
  @Contextual @SerialName("RECEIVE_TIME_FROM") val receiveTimeFrom: LocalTime,
  @Contextual @SerialName("RECEIVE_TIME_TO") val receiveTimeTo: LocalTime,
  @Contextual @SerialName("EDUCATE_FROM") val educateFrom: LocalDate,
  @Contextual @SerialName("EDUCATE_TO") val educateTo: LocalDate,
  @Contextual @SerialName("EDUCATE_TIME_FROM") val educateTimeFrom: LocalTime,
  @Contextual @SerialName("EDUCATE_TIME_TO") val educateTimeTo: LocalTime,
  @SerialName("MONDAY") val monday: String?,
  @SerialName("TUESDAY") val tuesday: String?,
  @SerialName("WEDNESDAY") val wednesday: String?,
  @SerialName("THURSDAY") val thursday: String?,
  @SerialName("FRIDAY") val friday: String?,
  @SerialName("SATURDAY") val saturday: String?,
  @SerialName("SUNDAY") val sunday: String?,
  @Serializable(with = DecimalIntSerializer::class) @SerialName("COLLECT_NUM") val collectNum: Int,
  @Serializable(with = DecimalIntSerializer::class) @SerialName("SPARE_NUM") val spareNum: Int,
  @Serializable(with = DecimalIntSerializer::class) @SerialName("EDUCATE_FEE") val educateFee: Int,
  @SerialName("VISIT_RECEIVE_FLAG") val visitReceiveFlag: String,
  @SerialName("ONLINE_RECEIVE_FLAG") val onlineReceiveFlag: String,
  @SerialName("URL") val url: String
) {
  /**
   * 목록 키: 센터(organCode)와 강좌 코드 조합
   */
  val key
    get() = "$organCode-$classCode"

  /**
   * 난이도
   */
  val difficulty
    get() = "[${difficultyName}]"

  /**
   * 신청 기간
   */
  val receivePeriod
    get() = if (receiveFrom == receiveTo) {
      "$receiveFrom(${receiveFrom.dayName}) $receiveTimeFrom ~ $receiveTimeTo"
    } else {
      "$receiveFrom(${receiveFrom.dayName}) $receiveTimeFrom ~\n$receiveTo(${receiveTo.dayName}) $receiveTimeTo"
    }

  /**
   * 교육 기간
   */
  val educatePeriod
    get() = if (educateFrom == educateTo) {
      "$educateFrom(${educateFrom.dayName})"
    } else {
      "$educateFrom(${educateFrom.dayName}) ~ $educateTo(${educateTo.dayName})"
    }

  /**
   * 교육 요일, 시간
   */
  val educateDays
    get() = if (educateFrom == educateTo) {
      educateTime
    } else {
      StringBuilder().apply {
        append("\n")
        if (!monday.isNullOrBlank()) append("월")
        if (!tuesday.isNullOrBlank()) append("화")
        if (!wednesday.isNullOrBlank()) append("수")
        if (!thursday.isNullOrBlank()) append("목")
        if (!friday.isNullOrBlank()) append("금")
        if (!saturday.isNullOrBlank()) append("토")
        if (!sunday.isNullOrBlank()) append("일")
        append("\n")
        append(educateTime)
      }.toString()
    }

  /**
   * 교육 시간
   */
  private val educateTime
    get() = "$educateTimeFrom ~ $educateTimeTo"

  /**
   * 잔여
   */
  val remainNumber
    get() = "$spareNum/${collectNum}명"

  /**
   * 수강료
   */
  val fee
    get() = if (educateFee > 0) {
      "%,d원".format(educateFee)
    } else {
      "무료"
    }

  /**
   * 접수 방법
   */
  val howToRegister
    get() = mutableListOf<String>().apply {
      if (visitReceiveFlag == "Y") add("방문")
      if (onlineReceiveFlag == "Y") add("온라인")
      if (listOf(visitReceiveFlag, onlineReceiveFlag).all { it == "N" }) add("접수방법 확인")
    }.joinToString()
}
